File: frontend/lib/features/knowledge_hub/domain/knowledge_hub.dart
class PdfFolder {
  const PdfFolder({
    required this.id,
    required this.name,
    required this.parentId,
  });

  final String id;
  final String name;
  final String? parentId;

  factory PdfFolder.fromJson(Map<String, dynamic> json) => PdfFolder(
    id: json['id'] as String,
    name: json['name'] as String,
    parentId: json['parent_id'] as String?,
  );
}

class PdfDocument {
  const PdfDocument({
    required this.id,
    required this.folderId,
    required this.filename,
    required this.sizeBytes,
    required this.pageCount,
    required this.updatedAt,
    required this.title,
    required this.author,
    required this.contentPreview,
  });

  final String id;
  final String? folderId;
  final String filename;
  final int sizeBytes;
  final int pageCount;
  final DateTime updatedAt;

  /// Extracted once at upload from the PDF's own embedded metadata / first
  /// page of text (see backend _extract_pdf_metadata) — null for scanned or
  /// image-only PDFs, or any PDF uploaded before this field existed.
  final String? title;
  final String? author;
  final String? contentPreview;

  /// What to actually show as the document's name — most PDFs' embedded
  /// title metadata is missing or junk (e.g. "Microsoft Word - doc1"), so
  /// only prefer it when it looks like a real title.
  String get displayTitle =>
      (title != null && title!.trim().length > 3) ? title! : filename;

  factory PdfDocument.fromJson(Map<String, dynamic> json) => PdfDocument(
    id: json['id'] as String,
    folderId: json['folder_id'] as String?,
    filename: json['filename'] as String,
    sizeBytes: json['size_bytes'] as int,
    pageCount: json['page_count'] as int,
    updatedAt: DateTime.parse(json['updated_at'] as String),
    title: json['title'] as String?,
    author: json['author'] as String?,
    contentPreview: json['content_preview'] as String?,
  );
}

/// A full-text content match from Azure AI Search (see backend
/// app/services/search.py) — distinct from filename matches, which
/// PdfDocument's own list-with-`q` already covers.
class PdfSearchHit {
  const PdfSearchHit({
    required this.pdfId,
    required this.filename,
    required this.pageNumber,
    required this.snippet,
  });

  final String pdfId;
  final String filename;
  final int? pageNumber;
  final String snippet;

  factory PdfSearchHit.fromJson(Map<String, dynamic> json) => PdfSearchHit(
    pdfId: json['pdf_id'] as String,
    filename: json['filename'] as String,
    pageNumber: json['page_number'] as int?,
    snippet: json['snippet'] as String,
  );
}

/// One entry of the document's embedded outline (table of contents),
/// flattened by the backend with a nesting [level] for indentation.
class PdfOutlineEntry {
  const PdfOutlineEntry({
    required this.title,
    required this.page,
    required this.level,
  });

  final String title;
  final int page;
  final int level;

  factory PdfOutlineEntry.fromJson(Map<String, dynamic> json) =>
      PdfOutlineEntry(
        title: json['title'] as String,
        page: json['page'] as int,
        level: json['level'] as int,
      );
}

class PdfBookmark {
  const PdfBookmark({
    required this.id,
    required this.pdfId,
    required this.pageNumber,
    required this.label,
  });

  final String id;
  final String pdfId;
  final int pageNumber;
  final String label;

  factory PdfBookmark.fromJson(Map<String, dynamic> json) => PdfBookmark(
    id: json['id'] as String,
    pdfId: json['pdf_id'] as String,
    pageNumber: json['page_number'] as int,
    label: json['label'] as String,
  );
}

class NormalizedRect {
  const NormalizedRect({
    required this.x,
    required this.y,
    required this.width,
    required this.height,
  });

  final double x;
  final double y;
  final double width;
  final double height;

  factory NormalizedRect.fromJson(Map<String, dynamic> json) => NormalizedRect(
    x: (json['x'] as num).toDouble(),
    y: (json['y'] as num).toDouble(),
    width: (json['width'] as num).toDouble(),
    height: (json['height'] as num).toDouble(),
  );

  Map<String, dynamic> toJson() => {
    'x': x,
    'y': y,
    'width': width,
    'height': height,
  };
}

/// MVP annotation scope: highlight and note only (OPEN_QUESTIONS.md) — the
/// legacy's freehand pen/shapes/stickers/image tools are cut.
class PdfAnnotation {
  const PdfAnnotation({
    required this.id,
    required this.pdfId,
    required this.pageNumber,
    required this.type,
    required this.color,
    required this.rects,
    required this.text,
    required this.note,
  });

  final String id;
  final String pdfId;
  final int pageNumber;
  final String type;
  final String color;
  final List<NormalizedRect> rects;
  final String text;
  final String note;

  factory PdfAnnotation.fromJson(Map<String, dynamic> json) => PdfAnnotation(
    id: json['id'] as String,
    pdfId: json['pdf_id'] as String,
    pageNumber: json['page_number'] as int,
    type: json['type'] as String,
    color: json['color'] as String,
    rects: (json['rects'] as List<dynamic>)
        .map((e) => NormalizedRect.fromJson(e as Map<String, dynamic>))
        .toList(),
    text: json['text'] as String,
    note: json['note'] as String,
  );
}

File: backend/app/models/knowledge_hub.py
import uuid
from datetime import datetime

from sqlalchemy import Boolean, DateTime, ForeignKey, Integer, String, Text, func
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


class PdfFolder(Base):
    __tablename__ = "pdf_folders"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False, index=True)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    parent_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("pdf_folders.id"), nullable=True)

    is_deleted: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)

    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())


class PdfDocument(Base):
    __tablename__ = "pdf_documents"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False, index=True)
    folder_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("pdf_folders.id"), nullable=True)

    filename: Mapped[str] = mapped_column(String(500), nullable=False)
    size_bytes: Mapped[int] = mapped_column(Integer, nullable=False)
    page_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)

    # Extracted once at upload time from the PDF's own embedded metadata /
    # first page of text (see _extract_pdf_metadata) — owner request,
    # 2026-08-14: the list showed nothing but filename/page count/size, no
    # indication of what's actually in the document. Null title/author means
    # the PDF simply doesn't carry that metadata (common for scans).
    title: Mapped[str | None] = mapped_column(String(500), nullable=True)
    author: Mapped[str | None] = mapped_column(String(255), nullable=True)
    content_preview: Mapped[str | None] = mapped_column(Text, nullable=True)

    is_deleted: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)

    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

    @property
    def storage_filename(self) -> str:
        # The on-disk name never changes even if the user renames the document —
        # `filename` is display-only, so renaming is a pure DB update.
        return f"{self.id}.pdf"


class PdfBookmark(Base):
    __tablename__ = "pdf_bookmarks"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    pdf_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("pdf_documents.id", ondelete="CASCADE"), nullable=False, index=True)
    page_number: Mapped[int] = mapped_column(Integer, nullable=False)
    label: Mapped[str] = mapped_column(String(255), nullable=False, default="")

    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())


class PdfAnnotation(Base):
    """MVP annotation scope: highlight and note only (OPEN_QUESTIONS.md) —
    the legacy's Konva freehand pen/shapes/stickers/image tools are cut,
    mirroring the freehand-drawing cut already made for Notes."""

    __tablename__ = "pdf_annotations"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    pdf_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("pdf_documents.id", ondelete="CASCADE"), nullable=False, index=True)
    page_number: Mapped[int] = mapped_column(Integer, nullable=False)
    type: Mapped[str] = mapped_column(String(20), nullable=False)  # "highlight" | "note"
    color: Mapped[str] = mapped_column(String(20), nullable=False, default="#FFEB3B")
    # Rects are page-relative fractions (0..1) so they stay correct regardless
    # of render resolution/zoom: [{"x","y","width","height"}, ...].
    rects: Mapped[list] = mapped_column(JSONB, nullable=False, default=list)
    text: Mapped[str] = mapped_column(Text, nullable=False, default="")
    note: Mapped[str] = mapped_column(Text, nullable=False, default="")

    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
